File: src/gateway/proxy/router.go
package proxy

import (
	"encoding/json"
	"fmt"

	"gateway/proxy/routerjs"

	"github.com/gorilla/mux"
	"github.com/robertkrimen/otto"
)

type jsRoute struct {
	Methods    []string          `json:"_methods"`
	Schemes    []string          `json:"_schemes"`
	Host       string            `json:"_host"`
	Path       string            `json:"_path"`
	PathPrefix string            `json:"_pathPrefix"`
	Headers    map[string]string `json:"_headers"`
	Queries    map[string]string `json:"_queries"`
	Name       string            `json:"_name"`
}

// ParseRoutes parses the script and stores the resulting the mux.Router.
func ParseRoutes(script *otto.Script) (*mux.Router, error) {
	vm := otto.New()

	var files = []string{"route.js", "router.js"}
	var scripts = make([]interface{}, 0)
	for _, filename := range files {
		fileJS, err := routerjs.Asset(filename)
		if err != nil {
			return nil, err
		}
		scripts = append(scripts, fileJS)
	}
	scripts = append(scripts, script)
	scripts = append(scripts, "router.routeData();")

	var result otto.Value
	for _, script := range scripts {
		var err error
		result, err = vm.Run(script)
		if err != nil {
			return nil, err
		}
	}

	var jsRoutes []jsRoute
	if err := json.Unmarshal([]byte(result.String()), &jsRoutes); err != nil {
		return nil, err
	}

	router := mux.NewRouter()
	for _, route := range jsRoutes {
		addRoute(route, router)
	}

	return router, nil
}

func addRoute(js jsRoute, router *mux.Router) error {
	route := router.NewRoute()

	if js.Name == "" {
		return fmt.Errorf("All routes must route to a proxy endpoint by name.")
	}
	route.Name(js.Name)

	if len(js.Methods) > 0 {
		route.Methods(js.Methods...)
	}

	if len(js.Schemes) > 0 {
		route.Schemes(js.Schemes...)
	}

	if js.Host != "" {
		route.Host(js.Host)
	}

	if js.Path != "" {
		route.Path(js.Path)
	}

	if js.PathPrefix != "" {
		route.PathPrefix(js.PathPrefix)
	}

	if len(js.Headers) > 0 {
		var pairs []string
		for k, v := range js.Headers {
			pairs = append(pairs, k, v)
		}
		route.Headers(pairs...)
	}

	if len(js.Queries) > 0 {
		var pairs []string
		for k, v := range js.Queries {
			pairs = append(pairs, k, v)
		}
		route.Queries(pairs...)
	}

	return nil
}

File: src/gateway/proxy/routerjs/static/router.js
var AP = AP || {};

AP.Router = function() {
  this.routes = [];
}

AP.Router.prototype.routeData = function() {
  return JSON.stringify(this.routes);
}

AP.Router.prototype.newRoute = function() {
  var route = new AP.Route();
  this.routes.push(route);
  return route;
}

AP.Router.prototype.httpMethod = function(method, path, name) {
  return this.newRoute().httpMethod(method, path, name);
}

AP.Router.prototype.get = function(path, name) {
  return this.newRoute().get(path, name);
}

AP.Router.prototype.post = function(path, name) {
  return this.newRoute().post(path, name);
}

AP.Router.prototype.put = function(path, name) {
  return this.newRoute().put(path, name);
}

AP.Router.prototype.patch = function(path, name) {
  return this.newRoute().patch(path, name);
}

AP.Router.prototype.delete = function(path, name) {
  return this.newRoute().delete(path, name);
}

AP.Router.prototype.method = function(method) {
  return this.newRoute().method(method);
}

AP.Router.prototype.methods = function() {
  var route = this.newRoute();
  return this.newRoute().methods.apply(route, arguments);
}

AP.Router.prototype.path = function(path) {
  return this.newRoute().path(path);
}

AP.Router.prototype.pathPrefix = function(pathPrefix) {
  return this.newRoute().pathPrefix(pathPrefix);
}

AP.Router.prototype.host = function(host) {
  return this.newRoute().host(host);
}

AP.Router.prototype.scheme = function(scheme) {
  return this.newRoute().scheme(scheme);
}

AP.Router.prototype.schemes = function() {
  var route = this.newRoute();
  return this.newRoute().schemes.apply(route, arguments);
}

AP.Router.prototype.headers = function(headers) {
  return this.newRoute().headers(headers);
}

AP.Router.prototype.queries = function(queries) {
  return this.newRoute().queries(queries);
}

AP.Router.prototype.name = function(name) {
  return this.newRoute().name(name);
}

var router = new AP.Router();
